Add a `B2Dataset` explicit wrapper

t20100 opened the pull request:
This PR propose to add a `B2Dataset` class that can be used to explicitly enable the access optimizations for a given h5py Dataset.

It builds on the same functions as the patching approach.

t20100 committed:
Add B2Dataset class

t20100 committed:
Add B2Dataset to example

t20100 committed:
add a small test

ivilata commented on b2h5py/blosc2.py:
```diff
@@ -189,3 +190,45 @@ def opt_slice_read(dataset, slice_, new_dtype=None):
+        """Whether or not Blosc2 fast slicing access is enabled"""
+        return opt_slicing_dataset_ok(self.dataset)
+
+    def __getitem__(self, args):
```
This can be much simplified if the `_blosc2_opt_slicing_ok` property is provided (equivalent to `is_b2_fast_access`), since the operation would be reduced to just attempting `opt_slice_read` (which does all the checking, also for `opt_slicing_enabled`, which is missing here) & catching `NoOptSlicingError` (within the `with` block), then calling the original getitem if it's caught. I.e. virtually what `b2h5py.patch.B2Dataset___getitem__` does. :wink:

t20100 replied:
Indeed, best to avoid copies.

ivilata commented on b2h5py/tests/test_b2dataset.py:
```diff
@@ -0,0 +1,37 @@
+import h5py
+import hdf5plugin
+import numpy as np
+import pytest
```
Please note that plain `b2h5py` tests don't use pytest, only vanilla `unittest` to avoid the extra dependency (which is only needed by the `h5py-test` extra feature to run h5py tests). You can see that this did cause a CI check to fail: <https://github.com/Blosc/b2h5py/actions/runs/7274572901/job/19860011744?pr=1>.

I think this should be rewritten for `unittest`.

t20100 replied:
Tests rewritten as a `TestCase`, CI passes

t20100 committed:
convert test to unittest

ivilata reviewed (changes requested):
Thanks @t20100 for your contribution, it looks very carefully developed and I mostly agree with integrating it, bar some issues (see comments):

- `B2Dataset.__getitem__` can be simplified and lacks some check.
- Tests should be converted to plain `unittest`.
- Finally, I suspect that this may have some issues with `__iter__`, as `Dataset.__iter__` calls back `__getitem__`, so it may drop the optimization chain there.

If you need some help or you prefer to leave some point up to me, please say so. Thanks!

t20100 commented on b2h5py/tests/test_b2dataset.py:
```diff
@@ -0,0 +1,37 @@
+    assert b2dataset.chunks == hdf5_test_dataset.chunks
+
+    # Access whole array
+    assert np.array_equal(b2dataset[()], hdf5_test_dataset[()])
```
As I implemented it, it silently fallback to `h5py.Dataset`... but that's indeed not consistent with the patched `h5py.Dataset`

t20100 commented:
> B2Dataset.__getitem__ can be simplified and lacks some check.

Indeed best to do as for the patched `dataset`

> I suspect that this may have some issues with __iter__, as Dataset.__iter__ calls back __getitem__, so it may drop the optimization chain there.

Yes, I didn't checked the proxying part. It's maybe best not to proxy `h5py.Dataset` methods/attrs with `__getattr__` and rather explicitly proxy a few useful information like `shape`, `ndim`, `dtype`, `chunks`, since it is accessible through to `dataset` attribute.

> If you need some help or you prefer to leave some point up to me, please say so. Thanks!

I won't be able to update it in the coming weeks so if you have some time and want to take it over in the mean time, it's all yours!

ivilata commented:
Thanks @t20100, I'll take it from here. :slightly_smiling_face:

ivilata committed:
Simplify `B2Dataset` test case by inheriting from h5py's.

ivilata committed:
Factor out common code for test storage and optimization checking.

Reuse in `B2Dataset` tests too.

ivilata committed:
Split out function to check if Blosc2 optimized slicing is possible.

ivilata committed:
Use optimization check function in `B2Dataset`.

This avoids repeating the checks, and also fixes the old code not checking
whether optimization is enabled globally.

ivilata committed:
Readme: Add mention to `B2Dataset` usage.

Also minor change to main package docstring for consistency.

ivilata committed:
Split Blosc2 optimization check into separate context manager.

ivilata committed:
Add note on required method for Blosc2 optimization check decorator.

ivilata committed:
Verify whether `B2Dataset` tests do use optimization when desired.

ivilata committed:
Add simple test for iteration on `B2Dataset` not hanging.

As `h5py.Dataset.__iter__` calls back `__getitem__`.

ivilata committed:
Use "fast slicing" instead of "fast access" regarding `B2Dataset`.

For consistency with the rest of code.

ivilata committed:
Avoid property access when proxying attributes in `B2Dataset`.

ivilata commented:
I think I fixed all the issues discussed above that were still pending. Regarding iteration, it does seem to work without issues, though it does indeed not use optimization. However, since iteration is not properly slicing, we're not breaking any interface promises anyway, so I just added a simple test to check that nothing breaks when iterating.

(I also had to change the approach to reusing `opt_*_read` functions in the end because some where expecting the monkey-patching, but that's just a fixed implementation detail.)

t20100 commented:
Many thanks @ivilata !
